src/gen/lib/Codegen_boilerplate.ml

```diff
@@ -0,0 +1,120 @@
+(*
+   Derive boilerplate to be used as a template when mapping a CST
+   to another type of tree.
+*)
+
+open Printf
+open CST_grammar
+open Indent.Types
+
+let trans = Codegen_util.translate_ident
+
+let make_header grammar = sprintf
+{|(**
+   Boilerplate to be used as a template when mapping the %s CST
+   to another type of tree.
+*)
+
+(* Disable warnings against unused variables *)
+[@@@warning "-26-27"]
+
+(* Disable warning against unused 'rec' *)
+[@@@warning "-39"]
+
+let token (_tok : Tree_sitter_run.Token.t) = failwith "not implemented"
+let blank () = failwith "not implemented"
+let todo _ = failwith "not implemented"
+
+|}
+    grammar.name
+
+let todo = [ Line "todo ()" ]
+
+let simple_pat = function
+  | Symbol _ -> "x"
+  | Token _ -> "tok"
+  | Blank -> "()"
+  | Repeat _ -> "xs"
+  | Repeat1 _ -> "xs"
+  | Choice _ -> "x"
+  | Optional _ -> "opt"
+  | Seq l ->
+      sprintf "(%s)"
+        (List.mapi (fun i _ -> sprintf "v%i" (i+1)) l
+         |> String.concat ", ")
+
+let simple_exp x = simple_pat x
+
+let gen_rule_mapper_body body =
+  let arg = simple_exp body in
+  match body with
+  | Symbol name -> [ Line (sprintf "map_%s %s" (trans name) arg) ]
+  | Token _token -> [ Line (sprintf "token %s" arg)]
+  | Blank -> [ Line (sprintf "blank %s" arg)]
+  | Repeat (Symbol name)
+  | Repeat1 (Symbol name) ->
+      [ Line (sprintf "List.map map_%s %s" (trans name) arg) ]
+  | Repeat (Token _token)
+  | Repeat1 (Token _token) ->
+      [ Line (sprintf "List.map token %s" arg) ]
+  | Repeat x
+  | Repeat1 x ->
+      [
+        Line (
+          sprintf "List.map (fun %s -> todo %s) %s"
+            (simple_pat x) (simple_exp x) arg
+        )
+      ]
+  | Choice l ->
+      let cases =
+        List.map (fun (name, body) ->
+          Line (sprintf "| `%s %s -> todo %s"
```

Comment:
I would actually in the body call recursively and then do the todo, for example

| `Cons (a,b,c) -> let a = stmt a in  let b = expr b in let c = token c in todo (a,b,c)

Reply:
so we factorize also the boilerplate of calling things recursively